### PR TITLE
When writing to STDOUT redirected to a closed pipe, no broken pipe

### PR DESCRIPTION
**Pull request by gogainda**
error message will be shown now

Ruby commit https://github.com/ruby/ruby/commit/6f28ebd585fba1aff1c9591ced08ed11b68ba9e3

before
```
~ % seq 100 | ruby -pe 'print if 12..' | head -2
1
2
<internal:core> core/errno.rb:54:in `handle_errno': Broken pipe (Errno::EPIPE)
	from <internal:core> core/posix.rb:449:in `write_string_native'
	from <internal:core> core/io.rb:2396:in `write'
	from <internal:core> core/kernel.rb:440:in `block in print'
	from <internal:core> core/kernel.rb:439:in `each'
	from <internal:core> core/kernel.rb:439:in `print'
	from -e:1:in `<main>'
~ % yes | ruby -pe 'print' | head -1
y
<internal:core> core/errno.rb:54:in `handle_errno': Broken pipe (Errno::EPIPE)
	from <internal:core> core/posix.rb:449:in `write_string_native'
	from <internal:core> core/io.rb:2396:in `write'
	from <internal:core> core/kernel.rb:440:in `block in print'
	from <internal:core> core/kernel.rb:439:in `each'
	from <internal:core> core/kernel.rb:439:in `print'
	from -e:1:in `<main>'
```

after
```
~ % seq 100 | ruby -pe 'print if 12..' | head -2
1
2
~ % yes | ruby -pe 'print' | head -1            
y
```

only problem I found is that -n is not working on truffleruby, so I don't know how it behaves with this fix

### CODE REVIEW

**Comment by gogainda** [comment]
apparently I broke other EPIPE related tests @eregon

**Comment by eregon** [comment]
This is part of the Ruby 3.0 changes:
> [medium] When writing to STDOUT redirected to a closed pipe, no broken pipe error message will be shown now. [Feature #14413]

File: src/main/ruby/truffleruby/core/posix.rb [hunk 1]
```diff
@@ -446,6 +446,7 @@ def self.write_string_native(io, string, continue_on_eagain)
               return written
             end
           else
+            raise SignalException.new(:SIGPIPE) if errno == 32
```
Comment (eregon) [inline src/main/ruby/truffleruby/core/posix.rb:449]: It should only be done for STDOUT, no other IO:
https://github.com/ruby/ruby/commit/6f28ebd585fba1aff1c9591ced08ed11b68ba9e3#diff-92194f057884b3287a3a6bf84e6e3b2bf433a556b68562799252a091744e7854R13438

And hardcoding 32 is problematic, use `Errno::EPIPE::Errno` instead.

**Comment by gogainda** [comment]
Not sure how to write test for it

**Comment by eregon** [comment]
@gogainda One way is to look at the CRuby test to see how it's tested. I'll write a spec for it this time.

**Review by eregon — APPROVED** [review]
Thank you!

**Comment by gogainda** [comment]
@eregon is there anything I can do to move this PR forward?

**Comment by eregon** [comment]
Sorry, I somehow lost sight of this PR, it looks like it didn't merge due to a new CI job being added and it didn't automatically run it.
I'll rebase and put it in the merge queue.

**Comment by eregon** [comment]
Actually there are two MRI tests failing:
```
  1) Failure:
TestPipe#test_stdout_epipe [/b/b/e/main/test/mri/tests/ruby/test_pipe.rb:40]:
[Errno::EPIPE] exception expected, not #<SignalException: SIGPIPE>.

  2) Failure:
TestPipe::WithConversion#test_stdout_epipe [/b/b/e/main/test/mri/tests/ruby/test_pipe.rb:40]:
[Errno::EPIPE] exception expected, not #<SignalException: SIGPIPE>.
```

**Comment by eregon** [comment]
I'll just tag that test for now, it seems a weird edge case with `STDOUT.reopen` clearing the flag or so.
